conn.go

```diff
@@ -136,6 +136,293 @@ func NewConnWith(conn net.Conn, config ConnConfig) *Conn {
 	return c
 }
 
+func writeHeader(w *bufio.Writer, clientID string, apiKey apiKey, apiVersion apiVersion, correlationID, size int32) {
+	h := requestHeader{
+		ApiKey:        int16(apiKey),
+		ApiVersion:    int16(apiVersion),
+		CorrelationID: correlationID,
+		ClientID:      clientID,
+	}
+	h.Size = h.size() - 4 + size
+
+	// write message
+	h.writeTo(w)
+}
+
+// describeGroups retrieves the specified groups
+//
+// See http://kafka.apache.org/protocol.html#The_Messages_DescribeGroups
+func (c *Conn) describeGroups(request describeGroupsRequestV1) (describeGroupsResponseV1, error) {
+	var response describeGroupsResponseV1
+
+	err := c.readOperation(
+		func(deadline time.Time, id int32) error {
+			w := &c.wbuf
+			writeHeader(w, c.clientID, describeGroupsRequest, v1, id, request.size())
+			request.writeTo(w)
+			return w.Flush()
+		},
+		func(deadline time.Time, size int) error {
+			return expectZeroSize(func() (remain int, err error) {
+				return (&response).readFrom(&c.rbuf, size)
+			}())
+		},
+	)
+	if err != nil {
+		return response, err
+	}
+	for _, group := range response.Groups {
+		if group.ErrorCode != 0 {
+			return response, Error(group.ErrorCode)
```

Comment:
Can we return `describeGroupsResponseV1{}` instead of `response` when there is a non-nil error? It seems less error-prone to return the zero-value when an error occurs instead of leaking values.

Reply:
I'm open to that.  The consideration was the loss of information.  If I return the zero value, you would love the knowledge of which group failed.  I'm happy to remove it for now since I'm not using the return value.

Reply:
I believe this line of reason will also apply to a number of the comments below.